File: hse/BPM161/bartosh/2/src/test/java/ru/spbau/gbarto/LockFreeSetTest.java
package ru.spbau.gbarto;

import org.junit.Test;

import java.util.Iterator;

import static org.junit.Assert.*;

public class LockFreeSetTest {
    @Test
    public void add() {
        LockFreeSetInterface<Integer> set = new LockFreeSet<>();

        assertTrue(set.add(8));
        assertTrue(set.add(800));
        assertTrue(set.add(555));
        assertTrue(set.add(3535));
        assertFalse(set.add(8));
        assertFalse(set.add(800));
        assertFalse(set.add(555));
        assertFalse(set.add(3535));
    }

    @Test
    public void remove() {
        LockFreeSetInterface<Integer> set = new LockFreeSet<>();

        set.add(8);
        set.add(800);

        assertTrue(set.remove(8));
        assertTrue(set.remove(800));
        assertFalse(set.remove(555));
        assertFalse(set.remove(3535));
    }

    @Test
    public void contains() {
        LockFreeSetInterface<Integer> set = new LockFreeSet<>();

        set.add(8);
        set.add(800);

        assertTrue(set.contains(8));
        assertTrue(set.contains(800));
        assertFalse(set.contains(555));
        assertFalse(set.contains(3535));
    }

    @Test
    public void isEmpty() {
        LockFreeSetInterface<Integer> set = new LockFreeSet<>();

        assertTrue(set.isEmpty());

        set.add(8);
        set.add(800);

        assertFalse(set.isEmpty());

        set.remove(8);
        set.remove(800);

        assertTrue(set.isEmpty());
    }

    @Test
    public void iterator() {
        LockFreeSetInterface<Integer> set = new LockFreeSet<>();

        set.add(8);
        set.add(800);
        set.add(555);
        set.add(3535);

        Iterator<Integer> setIterator = set.iterator();

        assertEquals(8, (int) setIterator.next());
        assertEquals(555, (int) setIterator.next());
        assertEquals(800, (int) setIterator.next());
        assertEquals(3535, (int) setIterator.next());
    }
}

File: hse/BPM161/fedotov/2/src/test/kotlin/ru/hse/iisuslik/LockFreeSetTest.kt
package ru.hse.iisuslik

import org.junit.Test

import org.junit.Assert.*

class LockFreeSetTest {

    @Test
    fun isEmpty() {
        val set = LockFreeSet<Int>()
        assertTrue(set.isEmpty())
    }

    @Test
    fun isEmptyAfterRemove() {
        val set = LockFreeSet<Int>()
        assertTrue(set.isEmpty())
        set.add(1)
        assertFalse(set.isEmpty())
        set.remove(1)
        assertTrue(set.isEmpty())
    }

    @Test
    fun add1() {
        val set = LockFreeSet<Int>()
        set.add(1)
        assertTrue(set.contains(1))
    }

    @Test
    fun add2() {
        val set = LockFreeSet<Int>()
        set.add(1)
        set.add(2)
        assertTrue(set.contains(1))
        assertTrue(set.contains(2))
    }

    @Test
    fun addAgain() {
        val set = LockFreeSet<Int>()
        assertTrue(set.add(1))
        assertTrue(set.add(2))
        assertFalse(set.add(2))
        assertTrue(set.contains(1))
        assertTrue(set.contains(2))
    }

    @Test
    fun removeFirst() {
        val set = LockFreeSet<Int>()
        set.add(1)
        set.add(2)
        set.add(3)
        set.remove(1)
        assertFalse(set.contains(1))
        assertTrue(set.contains(2))
        assertTrue(set.contains(3))
    }

    @Test
    fun removeLast() {
        val set = LockFreeSet<Int>()
        set.add(1)
        set.add(2)
        set.add(3)
        set.remove(3)
        assertTrue(set.contains(1))
        assertTrue(set.contains(2))
        assertFalse(set.contains(3))
    }

    @Test
    fun removeMiddle() {
        val set = LockFreeSet<Int>()
        set.add(1)
        set.add(2)
        set.add(3)
        set.remove(2)
        assertTrue(set.contains(1))
        assertFalse(set.contains(2))
        assertTrue(set.contains(3))
    }


    @Test
    fun iterator() {
        val set = LockFreeSet<Int>()
        set.add(1)
        set.add(2)
        set.add(3)
        val it = set.iterator()
        val list = mutableListOf<Int>()
        it.forEach { list.add(it) }
        assertEquals(listOf(1, 2, 3), list)
    }

    @Test
    fun iteratorWith2Adds() {
        val set = LockFreeSet<Int>()
        set.add(1)
        set.add(2)
        set.add(2)
        set.add(3)
        val it = set.iterator()
        val list = mutableListOf<Int>()
        it.forEach { list.add(it) }
        assertEquals(listOf(1, 2, 3), list)
    }
}